File: CSM_Exemple_6_PD_2n_ordre.m
%%% Exemple sintonització analítica PD de 2n ordre
% Identificar el model, el diagrama de Bode, i implementar PI
% J.A.Mayugo, UdG, 2022

clear; close all;

%% Propietats del sistema
s = tf('s');
G_m = 150/((1+0.15*s)*s);

e_ss = 0;   % eu, error en règim estacionari nul
SP = 5;     % %,  valor del sobre pic (SP) inferior a 5%
ts2 = 0.57; % s,  temps per error inferior al 2%

U = 1;
t_ = 0:0.01:1;

% Resposta en llaç obert:
U_ = U *ones(size(t_));
[omg,T,X]         = lsim(G_m,U_,t_);  % voltage/angular vel.
H_ol=figure;
plot(t_,omg,'LineWidth',1.8)
grid; title(['Resposta a una funció escaló amb ' num2str(U) ' V al motor'],'FontSize',24)
xlabel('$t$ [s]','Interpreter','latex','FontSize',24);ylabel('$\omega$ [rad/s]','Interpreter','latex','FontSize',24)
set(gca,'FontSize',18)
saveas(H_ol,'./figures/Exemple_PD_2n_ordre_ol.pdf')

% FUNCIÓ AMB L'ORDRE BODE de MatLab i R-Locus
H_bode=figure
h = bodeplot(G_m); %,{0.1,20});
%setoptions(h,'MagUnits','abs','FreqScale','linear')

H_rlocus=figure;
rlocus(G_m);
p = pole(G_m)

%% APARTA b) i c) Sintonització analítica d'un PD
% control velocitat llaç tancat

A = ((log(100) - log(SP))/pi)^2
zeta = (A/(1+A))^0.5   % factor d'esmorteïment
omega_n = 4/(zeta*ts2) % rad/s

Kp = omega_n^2/1000 
Ki = 0
Kd = (2*zeta*omega_n-10)/1000

%Kp = 0.1034
%Kd = 0.0040
PID = Kp + Ki/s + Kd*s        % controlador PID
G_cl = PID*G_m/(1+PID*G_m)    % model llaç tancat amb PID, fórmula de Mason
p_cl = pole(G_cl)
z_cl = zero(G_cl)

E = 1/(1+PID*G_m)

% Model de segon ordre pur
G_cli = 1000*Kp/(s^2+(10+1000*Kd)*s+1000*Kp)
%G_cli = PID_G_mi/(1+PID_G_mi)  % model llaç tancat amb PID, fórmula de Mason
p_cli = pole(G_cli)
z_cli = zero(G_cli)

u = 1;
u_ = u *ones(size(t_));
u05_ = 1.05*u_;
u02_ = 1.02*u_;

[omg_,T,X]         = lsim(G_cl,u_,t_);  % voltage/angular vel.
[omgi_,T,X]        = lsim(G_cli,u_,t_);  % voltage/angular vel.
H_cl=figure;hold on;
plot(t_,omg_,'LineWidth',1.8)
plot(t_,omgi_,'LineWidth',1.8)
plot(t_,u02_)
plot(t_,u05_)
plot([ts2 ts2],[0 1.02])
grid; title(['Resposta control PD, consigna ' num2str(u) ' ue'],'FontSize',24)
xlabel('$t$ [s]','Interpreter','latex','FontSize',24);ylabel('$UE$ [ue]','Interpreter','latex','FontSize',24)
set(gca,'FontSize',18)
saveas(H_cl,'./figures/Exemple_PD_2n_ordre_cl.pdf')

File: CSM_Exemple_6_PD_2n_ordre.py
# Exemple de sintonització analítica d'un PD de 2n ordre
# Sense la llibreria "control"
# Implementació amb scipy.signal

import numpy as np
import matplotlib.pyplot as plt
from scipy import signal

# ---------------------------------------------------
# Propietats del sistema
# ---------------------------------------------------

SP = 5          # %
ts2 = 0.57      # s

t_ = np.arange(0, 1, 0.01)

# ---------------------------------------------------
# MODEL DEL MOTOR
#
# Gm(s) = 100 / ((1 + 0.1 s) s)
#
# Denominador:
# (1 + 0.1 s)s = 0.1 s² + s
# ---------------------------------------------------

num_Gm = [100]
den_Gm = [0.1, 1, 0]

Gm = signal.TransferFunction(num_Gm, den_Gm)

# ---------------------------------------------------
# Resposta a esglaó en llaç obert
# ---------------------------------------------------

t_ol, y_ol = signal.step(Gm, T=t_)

plt.figure(figsize=(8,5))
plt.plot(t_ol, y_ol, linewidth=2)

plt.grid(True)
plt.title("Resposta a esglaó en llaç obert")
plt.xlabel("t [s]")
plt.ylabel("ω [rad/s]")

plt.show()

# ---------------------------------------------------
# Sintonització PD
# ---------------------------------------------------

A = ((np.log(100) - np.log(SP)) / np.pi)**2

zeta = np.sqrt(A / (1 + A))

omega_n = 4 / (zeta * ts2)

Kp = omega_n**2 / 1000
Kd = (2 * zeta * omega_n - 10) / 1000

print(f"ζ = {zeta:.4f}")
print(f"ω_n = {omega_n:.4f}")

print(f"Kp = {Kp:.6f}")
print(f"Kd = {Kd:.6f}")

# ---------------------------------------------------
# Controlador PD
#
# C(s) = Kp + Kd s
# ---------------------------------------------------

num_PD = [Kd, Kp]
den_PD = [1]

# ---------------------------------------------------
# Funció de transferència en llaç obert:
#
# L(s) = C(s) G(s)
# ---------------------------------------------------

num_ol = np.polymul(num_PD, num_Gm)
den_ol = np.polymul(den_PD, den_Gm)

# ---------------------------------------------------
# Llaç tancat:
#
# Gcl = L(s)/(1+L(s))
# ---------------------------------------------------

num_cl = num_ol

den_cl = np.polyadd(den_ol, num_ol)

Gcl = signal.TransferFunction(num_cl, den_cl)

# ---------------------------------------------------
# Model ideal de segon ordre
# ---------------------------------------------------

num_cli = [1000 * Kp]

den_cli = [
    1,
    (10 + 1000 * Kd),
    1000 * Kp
]

Gcli = signal.TransferFunction(num_cli, den_cli)

# ---------------------------------------------------
# Resposta temporal
# ---------------------------------------------------

t_cl, y_cl = signal.step(Gcl, T=t_)

t_cli, y_cli = signal.step(Gcli, T=t_)

# Línies auxiliars
u02 = 1.02 * np.ones_like(t_)
u05 = 1.05 * np.ones_like(t_)

# ---------------------------------------------------
# Gràfica comparativa
# ---------------------------------------------------

plt.figure(figsize=(10,6))

plt.plot(t_cl, y_cl, linewidth=2, label='Sistema real PD')

plt.plot(t_cli, y_cli, linewidth=2, label='Model 2n ordre')

plt.plot(t_, u02, 'k--', label='2%')

plt.plot(t_, u05, 'r--', label='5%')

plt.plot([ts2, ts2], [0, 1.02], 'g--')

plt.grid(True)

plt.title("Resposta control PD")

plt.xlabel("t [s]")

plt.ylabel("UE [ue]")

plt.legend()

plt.show()